### PR TITLE
[NVIDIA XLA] Support Conv-Bias-Elu fusion in XLA using cuDNN runtime fusion

### PR DESCRIPTION
**Pull request by kaixih**
This PR enables the Conv-Bias-Elu fusion in XLA, which can take advantage of the cuDNN runtime compiled fusion kernels.

XLA users need to set `xla_gpu_use_runtime_fusion` to enable this behavior as we use `TF_CUDNN_USE_RUNTIME_FUSION` in the native TF.

cc. @nluehr @pjannaty

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc [hunk 1]
```diff
@@ -445,6 +454,58 @@ StatusOr<bool> FuseSideInputAlpha(HloComputation* comp) {
...
+      continue;
+    }
+    TF_ASSIGN_OR_RETURN(CudnnConvBackendConfig config,
+                        conv->backend_config<CudnnConvBackendConfig>());
+    if (config.activation_mode() != se::dnn::kNone) {
+      continue;
+    }
+
+    if (!ConsumeFuel("cudnn-fused-convolution-rewriter", [&] {
+          return absl::StrCat("FuseRelu: ", conv->ToString());
```
Comment (SandSnip3r) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc:496]: Typo in string, FuseElu

**Comment by cheshire** [comment]
Is there a reason not to enable this by default?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc [hunk 2]
```diff
@@ -764,6 +764,113 @@ TEST_F(CudnnFusedConvRewriterHloTest, DontFuseReluIfMultipleUses) {
...
+                    gte_pattern,
+                    m::Op()
+                        .WithPredicate([](const HloInstruction* instr) {
+                          return instr->opcode() == HloOpcode::kExpm1;
+                        })
+                        .WithOperand(0, gte_pattern)),
+          m::Minimum())));
+  TF_ASSERT_OK_AND_ASSIGN(auto config,
+                          conv->backend_config<CudnnConvBackendConfig>());
+  EXPECT_EQ(config.activation_mode(), se::dnn::kNone);
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:871]: I think we would need an execution test as well?

File: tensorflow/compiler/xla/xla.proto [hunk 3]
```diff
@@ -424,7 +424,11 @@ message DebugOptions {
   // (much) faster on our hardware.  Set this flag to disable this behavior.
   bool xla_cpu_strict_dot_conv_math = 175;
 
-  // Next id: 179
+  // An option to enable using cuDNN runtime compiled fusion kernels which is
+  // available and recommended for Ampere+ GPUs.
+  bool xla_gpu_use_runtime_fusion = 179;
```
Comment (cheshire) [inline tensorflow/compiler/xla/xla.proto:429]: Probably better to explicitly set to false ? Do you intend to make it true by default?

**Review by cheshire — CHANGES_REQUESTED** [review]

**Comment by kaixih** [comment]
> Is there a reason not to enable this by default?

Yes, the runtime compilation usually comes with overhead. For example, for the Ampere GPUs, the overhead is typically within 2ms for every engine. So, I disabled them for native TF by default, mainly because in the native TF, for every autotuning step, we need to query the heuristics (needing runtime compilation) and sweep them to pick the currently best one. In other words, every autotuning step will see the compilation overhead.

For XLA, the story is different because IIRC in XLA we can query the heuristics once (needing runtime compilation) and autotune them multiple times (no complication). But to be more conservative, I also disabled them in this PR. Of cause, I am ok to enable them by default, but I really want to hear from you (XLA): is it ok to introduce such overhead in the XLA scenario? If not, is that because the overhead is too much? And what is the allowable/expected overhead? Our cuDNN team keeps working on further reduction of the overhead but wanted to know the target. Thanks.

**Comment by cheshire** [comment]
We can try to run our benchmarks first, then it's probably simplest to enable it by default in this PR. Having it disabled by default is not great, as we'll essentially get a lot of dead untested code.

**Comment by kaixih** [comment]
Sure I will do that. Thanks for the feedback. Also there is a typo the above comment: `within 2ms -> within 2s`. Fixed the original comment.

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc (hunk 2 above)
Comment (kaixih) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:871]: Can you point me to which test that I can follow for this execution test? Thanks.

**Comment by cheshire** [comment]
Seems to break `//tensorflow/compiler/xla/service/gpu:cudnn_fused_conv_rewriter_test` when running with  `XLA_FLAGS=--xla_gpu_enable_xla_runtime_executable`, judging by the error message the error is with the test. 

@ezhulenev

**Comment by kaixih** [comment]
There is another reason why I didn't enable it by default. This feature is Ampere-ready, so it is recommended for Ampere+ GPUs, otherwise the compilation time could be very slow (e.g., reaching ~20s on Volta). That was why in the [PR](https://github.com/tensorflow/tensorflow/pull/56572) for the native TF, we will check the GPU types during the remapping even when users have set `TF_CUDNN_USE_RUNTIME_FUSION`. I am not sure it is a typical thing in XLA to check GPU types in the rewriting phase (Any sample codes?). If yes, this feature can be set to be on by default. @cheshire

**Comment by kaixih** [comment]
I have added the execution test and made the feature on by default. The rewriter will check the underlying GPU generation to determine if the runtime fusion will be actually used or not.

PTAL. @cheshire

**Comment by cheshire** [comment]
> therwise the compilation time could be very slow (e.g., reaching ~20s on Volta).

Checking compute capability before doing the rewrite makes sense to me.

**Comment by gcforster** [comment]
@kaixih Our internal benchmarking doesn't indicate any conclusive speedup (or slowdown) since it's noisy. It's preferred to have the fusion capability, so the plan is to move forward with the PR. Do you have test results that demonstrate speedup with this PR?

**Comment by kaixih** [comment]
Just pushed a new commit to make sure the inputs are 32-bit aligned. This is a requirement for the runtime fusion kernels. (Fyi, we have already done such checks in the remapper [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/remapper.cc#L520-L528).)

Re the benchmarks, I am syncing with our cudnn team to see if they have some results at hand that we can share. Otherwise, I will do some quick benchmarking. @gcforster

**Comment by cheshire** [comment]
> I've also collected some benchmark results by using the script [here](https://github.com/kaixih/dl_samples/blob/main/fusion/run_benchmark.sh). And the results can be found [here](https://docs.google.com/spreadsheets/d/187SnyG4wpwHNOmBa6MXjZQGHbB4sNAfhBFo82fHGxo8/edit?usp=sharing). Basically, about 5% to 40% perf gains can be seen. @gcforster

Ideally, could you post to open-xla forums with these results? Note that the current access to the sheet is restricted, maybe a GH gist is easier?

**Comment by cheshire** [comment]
> I've also collected some benchmark results by using the script [here](https://github.com/kaixih/dl_samples/blob/main/fusion/run_benchmark.sh). 

Those are synthetic benchmarks, I would be curious what is the gain on anything more real (as an example: fusing an elementwise operation after a matmul will give a large perf boost on a synthetic benchmark, but will probably make almost no difference on a real one, as the elementwise op will most likely be fused with subsequent ops in any case).

**Comment by kaixih** [comment]
> > I've also collected some benchmark results by using the script [here](https://github.com/kaixih/dl_samples/blob/main/fusion/run_benchmark.sh). And the results can be found [here](https://docs.google.com/spreadsheets/d/187SnyG4wpwHNOmBa6MXjZQGHbB4sNAfhBFo82fHGxo8/edit?usp=sharing). Basically, about 5% to 40% perf gains can be seen. @gcforster
> 
> Ideally, could you post to open-xla forums with these results? Note that the current access to the sheet is restricted, maybe a GH gist is easier?

Sure. This [link ](https://docs.google.com/spreadsheets/d/187SnyG4wpwHNOmBa6MXjZQGHbB4sNAfhBFo82fHGxo8/edit?usp=sharing) should be accessible. Will do the open-xla post later.

**Comment by kaixih** [comment]
I have just posted a new thread in the openxla forum: [here](https://github.com/openxla/xla/discussions/19).

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
Current test failure:

```
Expected: an HloInstruction:
 * with opcode get-tuple-element AND
 * with operand 0 which is:
     an HloInstruction:
      * with opcode custom-call AND
      * with 3 operands AND
      * with operand 0 which is:
          an HloInstruction:
           * with opcode parameter AND
           * which is parameter 0 AND
      * with operand 1 which is:
          an HloInstruction:
           * with opcode parameter AND
           * which is parameter 1 AND
      * with operand 2 which is:
          an HloInstruction:
           * with opcode parameter AND
           * which is parameter 2 AND
      * custom call with target '__cudnn$convBiasActivationForward' AND
 * which is a GTE with index 0 AND
 * outputting
     a shape:
      * with element type F16 AND
      * with dimensions [1,32,9,9]
  Actual: 0x36e97ea59880 (of type xla::HloInstruction*), HloInstruction doesn't have opcode get-tuple-element
in elu = f16[1,32,9,9]{3,2,1,0} select(pred[1,32,9,9]{3,2,1,0} cmp, f16[1,32,9,9]{3,2,1,0} get-tuple-element.1, f16[1,32,9,9]{3,2,1,0} expm1)

//tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:836: HloModule Test, entry_computation_layout={(f16[1,16,9,9]{3,2,1,0},f16[3,3,16,32]{3,2,1,0},f16[32]{0})->f16[1,32,9,9]{3,2,1,0}}

ENTRY %Test (inputs: f16[1,16,9,9], filters: f16[3,3,16,32], bias: f16[32]) -> f16[1,32,9,9] {
  %inputs = f16[1,16,9,9]{3,2,1,0} parameter(0)
  %filters = f16[3,3,16,32]{3,2,1,0} parameter(1)
  %bias = f16[32]{0} parameter(2)
  %cudnn-conv-bias-activation.2 = (f16[1,32,9,9]{3,2,1,0}, u8[0]{0}) custom-call(f16[1,16,9,9]{3,2,1,0} %inputs, f16[3,3,16,32]{3,2,1,0} %filters, f16[32]{0} %bias), window={size=3x3 pad=1_1x1_1}, dim_labels=bf01_01io->bf01, custom_call_target="__cudnn$convBiasActivationForward", backend_config="{\"activation_mode\":\"0\",\"conv_result_scale\":1,\"side_input_scale\":0}"
  %get-tuple-element.1 = f16[1,32,9,9]{3,2,1,0} get-tuple-element((f16[1,32,9,9]{3,2,1,0}, u8[0]{0}) %cudnn-conv-bias-activation.2), index=0
  %zero = f16[] constant(0)
  %zeros = f16[1,32,9,9]{3,2,1,0} broadcast(f16[] %zero), dimensions={}
  %cmp = pred[1,32,9,9]{3,2,1,0} compare(f16[1,32,9,9]{3,2,1,0} %get-tuple-element.1, f16[1,32,9,9]{3,2,1,0} %zeros), direction=GT
  %expm1 = f16[1,32,9,9]{3,2,1,0} exponential-minus-one(f16[1,32,9,9]{3,2,1,0} %get-tuple-element.1)
  ROOT %elu = f16[1,32,9,9]{3,2,1,0} select(pred[1,32,9,9]{3,2,1,0} %cmp, f16[1,32,9,9]{3,2,1,0} %get-tuple-element.1, f16[1,32,9,9]{3,2,1,0} %expm1)
}
```

No idea why is it not failing externally.. Maybe a different GPU used?

**Comment by kaixih** [comment]
I just force-pushed the commits to resolve the conflict.

@cheshire Thanks for the feedback of the failed test case. I just rebuilt the tests and didn't see any error. The following is what I used to test on my A6000 machine. Not sure if I missed anything. Can you provide more info on the failed tests, maybe like GPU?
```
$ bazel build --java_runtime_version=remotejdk_11 //tensorflow/compiler/xla/service/gpu:cudnn_fused_conv_rewriter_test
$ bazel-bin/tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test --gtest_filter=*.*Elu*
Note: Google Test filter = *.*Elu*
[==========] Running 3 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 1 test from CudnnFusedConvRewriterTest
[ RUN      ] CudnnFusedConvRewriterTest.TestElu
2022-10-25 20:14:21.735950: I tensorflow/compiler/xla/service/platform_util.cc:72] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- was support for that platform linked in?
2022-10-25 20:14:24.906099: I tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:428] Loaded cuDNN version 8600
2022-10-25 20:14:36.428248: W tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc:728] None of the algorithms provided by cuDNN heuristics worked; trying fallback algorithms.  Conv: (f64[1,3,3,64]{2,1,3,0}, u8[0]{0}) custom-call(f64[1,3,3,64]{2,1,3,0}, f64[3,3,64,64]{1,0,2,3}, f64[64]{0}), window={size=3x3 pad=1_1x1_1}, dim_labels=b01f_01io->b01f, custom_call_target="__cudnn$convBiasActivationForward", backend_config="{\"conv_result_scale\":1,\"activation_mode\":\"0\",\"side_input_scale\":0}"
[       OK ] CudnnFusedConvRewriterTest.TestElu (14825 ms)
[----------] 1 test from CudnnFusedConvRewriterTest (14826 ms total)

[----------] 2 tests from CudnnFusedConvRewriterHloTest
[ RUN      ] CudnnFusedConvRewriterHloTest.FuseElu
2022-10-25 20:14:36.561173: I tensorflow/compiler/xla/service/platform_util.cc:72] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- was support for that platform linked in?
[       OK ] CudnnFusedConvRewriterHloTest.FuseElu (4 ms)
[ RUN      ] CudnnFusedConvRewriterHloTest.DontFuseEluIfMultipleUses
2022-10-25 20:14:36.565504: I tensorflow/compiler/xla/service/platform_util.cc:72] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- was support for that platform linked in?
[       OK ] CudnnFusedConvRewriterHloTest.DontFuseEluIfMultipleUses (4 ms)
[----------] 2 tests from CudnnFusedConvRewriterHloTest (8 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 2 test suites ran. (14834 ms total)
[  PASSED  ] 3 tests.
```

**Comment by kaixih** [comment]
I can repro the failure on my V100 machine. Basically, we need to skip the `CudnnFusedConvRewriterHloTest.FuseElu` test on the pre-Ampere GPUs as well, because the test will expect the fused hlo instruction but the `RunHloPass` will check the GPU generation and skip the fusion when it is lower than Ampere. So, we should skip this test for the pre-Ampere platforms.

The fix has been submitted. PTAL. @cheshire

**Review by cheshire — APPROVED** [review]

**Comment by kaixih** [comment]
Gentle ping. @cheshire (It seems the "feedback/copybara" failed but I cannot access the log. Can you help?)

**Comment by cheshire** [comment]
Hm seems like `tensorflow/compiler/xla/service/gpu:cudnn_fused_conv_rewriter_test` is still failing =/

It fails comparing module equality in `CudnnFusedConvRewriterHloTest.DontFuseSideInputThroughRelu`

**Comment by kaixih** [comment]
Thanks @cheshire . Yes, I can reproduce the failure you mentioned. After some digging, I think I missed this statement for FuseElu.
```c++
any_changed |= changed;
```
Basically, every `TF_ASSIGN_OR_RETURN(changed, FuseXXX(comp))` needs its own one. PTAL the latest commit.

**Comment by gbaned** [comment]
@kaixih Can you please resolve conflicts? Thank you!

**Comment by kaixih** [comment]
Rebased to resolve the conflict. @cheshire

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc (hunk 2 above)
Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:906]: Any test which has RunAndCompare inside? Could make sense to have it in a separate file.

Comment (kaixih) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:906]: Yes, I have already followed your suggestion and added the test in
```c++
TEST_F(CudnnFusedConvRewriterTest, TestElu) {
......
```

**Review by cheshire — APPROVED** [review]

**Comment by thomasjoerg** [comment]
I will look into the failing CI build and get back to you.

**Comment by thomasjoerg** [comment]
A fix for the ARM CI build failure is underway.

@kaixih please resolve the conflicts meanwhile.

**Comment by kaixih** [comment]
Just rebased the code to resolve the conflict. PTAL @thomasjoerg. Thanks.

**Pull request closed without merging**